workers/unity/Packages/io.improbable.worker.sdk/IOStream.cs

```diff
@@ -0,0 +1,133 @@
+        {
+            stream.Dispose();
+        }
+
+        public static IOStream CreateRingBufferStream(uint capacity)
+        {
+            return new IOStream(CIO.CreateRingBufferStream(capacity));
+        }
+
+        public static IOStream CreateFileStream(string fileName, OpenMode openMode)
+        {
+            fixed (byte* fileNameBytes = ApiInterop.ToUtf8Cstr(fileName))
+            {
+                return new IOStream(CIO.CreateFileStream(fileNameBytes, (CIO.OpenMode) openMode));
+            }
+        }
+
+        public long Write(byte[] data)
+        {
+            var remainingCapacity = CIO.StreamGetRemainingWriteCapacityBytes(stream);
+            if (remainingCapacity < data.Length)
+            {
+                throw new IOException("Not enough stream capacity to write data.");
+            }
+
+            var bytesWritten = 0L;
+            fixed (byte* dataToWrite = data)
+            {
+                bytesWritten = CIO.StreamWrite(stream, dataToWrite, 1);
+            }
+
+            if (bytesWritten != -1)
+            {
+                return bytesWritten;
+            }
+
+            var rawError = CIO.StreamGetLastError(stream);
+            throw new IOException(ApiInterop.FromUtf8Cstr(rawError));
+        }
+
+        public long Read(uint bytesToRead, out byte[] streamData)
+        {
+            streamData = new byte[bytesToRead];
+
+            var bytesRead = 0L;
+            fixed (byte* streamDataPointer = streamData)
+            {
+                bytesRead = CIO.StreamRead(stream, streamDataPointer, bytesToRead);
+            }
+
+            if (bytesRead != -1)
+            {
+                return bytesRead;
+            }
+
+            var rawError = CIO.StreamGetLastError(stream);
+            throw new IOException(ApiInterop.FromUtf8Cstr(rawError));
+        }
+
+        public long Read(byte[] streamData)
+        {
+            var bytesToRead = (uint) streamData.Length;
+            var bytesRead = 0L;
+            fixed (byte* streamDataPointer = streamData)
+            {
+                bytesRead = CIO.StreamRead(stream, streamDataPointer, bytesToRead);
+            }
+
+            if (bytesRead != -1)
+            {
+                return bytesRead;
+            }
+
+            var rawError = CIO.StreamGetLastError(stream);
+            throw new IOException(ApiInterop.FromUtf8Cstr(rawError));
+        }
+
+        public byte[] Peek(uint bytes)
+        {
+            var streamData = new byte[bytes];
+
+            var bytesPeeked = 0L;
+            fixed (byte* streamDataPointer = streamData)
+            {
+                bytesPeeked = CIO.StreamPeek(stream, streamDataPointer, bytes);
+            }
+
+            if (bytesPeeked != -1)
+            {
+                return streamData;
+            }
+
+            var rawError = CIO.StreamGetLastError(stream);
+            throw new IOException(ApiInterop.FromUtf8Cstr(rawError));
+        }
```

Comment:
Should we make this `public uint Peek(uint bytes, out byte[] bytes)` to match the `Read` method?

Reply:
The `Read` method has two implementations, one where the caller provides the `bytes` array which is written to, and the other takes a the number of bytes to read and creates the array for them.

Should I add another implementation of `Peek` so that it matches `Read`?